### ISSUE TITLE
Setting SINGULARITY_CACHEDIR leads to missing singulairty images

### ISSUE THREAD

**Issue by openpaul** [nextflow-io/nextflow#1659]
### Expected behavior and actual behavior

**Expected:** Setting `SINGULARITY_CACHEDIR` in the .bashrc (.zshrc) leads to singularity images being downloaded and used from the folder specified there. 

**Actual behaviour:** Instead singularity images are downloaded to `SINGULARITY_CACHEDIR` as intended, but then nextflow (or singularity?) looks in the default directory for them and can not find them.

### Steps to reproduce the problem

Set `SINGULARITY_CACHEDIR` to any folder besides the global default and launch a script with a novel, not cached singularity container.

I can provide a minimal not working example if needed. But of course singularity is required.

### Program output 
```
ERROR  : Image path [...].img doesn't exist: No such file or directory
ABORT  : Retval = 255
```

### Environment 

* Nextflow version: 20.01.0.5264
* Java version: openjdk version "1.8.0_161"
* Operating system: CentOS
* Bash version: zsh 5.7.1

**Comment by zichner** [comment]
Part of the issue might be that there can be two different (types of) cache directories. The one of Singularity (defined by `SINGULARITY_CACHEDIR`), which contains downloaded image layers and much more, and the one of Nextflow (defined by the Nextflow variable `singularity.cacheDir`), which contains previously downloaded Singularity images.

So, you can try to leave `SINGULARITY_CACHEDIR` unchanged and only set `singularity.cacheDir`. You can even disable the Singularity caching of individual layers, etc by setting `SINGULARITY_DISABLE_CACHE=1` and only cache the downloaded images.

**Comment by openpaul** [comment]
Oh, that I did not know. Thank you.
Naive question, but shouldn't nextflow populate `singularity.cacheDir` with `SINGULARITY_CACHEDIR` in a perfect world?

**Comment by pditommaso** [comment]
I suspect this is a singularity issue. You can provide a test case and the singularity version?

**Comment by tskir** [comment]
I can confirm that this bug exists and manifests itself when:
1. `SINGULARITY_CACHEDIR` is set, **and**
2. `NXF_SINGULARITY_CACHEDIR` is either not set **or** set to a different location than the first variable.

### Test case to reproduce
```bash
mkdir A B
export SINGULARITY_CACHEDIR=`realpath A`
export NXF_SINGULARITY_CACHEDIR=`realpath B`
wget -q https://raw.githubusercontent.com/CINECA-project/wp4-federated-joint-cohort-analysis/4fbefeaf9e1570b25f54eafc805a5e39b9e6db34/4.3-pipelines/demonstrators/hello-world/hello_world.nf
nextflow run -with-singularity alpine:latest hello_world.nf
```

### Output
```
N E X T F L O W  ~  version 20.10.0
Launching `hello_world.nf` [jovial_laplace] - revision: 615d3a30a4
[-        ] process > say_hello -
Pulling Singularity image docker://alpine:latest [cache /homes/ktsukanov/test/B/alpine-latest.img]
Error executing process > 'say_hello'

Caused by:
  java.nio.file.NoSuchFileException: /homes/ktsukanov/test/B/alpine-latest.img.pulling.1613536027098
```

### Resulting file structure
Nextflow wants to pull the image into folder B, but in reality it ends up in folder A:
```
.
├── A
│   ├── alpine-latest.img.pulling.1613536027098
│   ├── docker
│   │   └── sha256:4c0d98bf9879488e0407f897d9dd4bf758555a78e39675e72b5124ccf12c2580.tar.gz
│   └── metadata
│       └── sha256:1f6d219d0077f6af480398f264117d2e74c03ec5f12d09b7b29b1eefbba872eb.tar.gz
├── B
├── hello_world.nf
├── nextflow
├── .nextflow
│   ├── cache
│   │   └── b28ef97e-6c38-4f08-974e-c3e8105eca70
│   │       ├── db
│   │       │   ├── 000003.log
│   │       │   ├── CURRENT
│   │       │   ├── LOCK
│   │       │   └── MANIFEST-000002
│   │       └── index.jovial_laplace
│   └── history
├── .nextflow.log
└── work
```

A similar thing happens when `NXF_SINGULARITY_CACHEDIR` in the example above is unset, but the rest is the same.

### Versions used
```
$ singularity --version
2.6.0-dist
$ nextflow -v
nextflow version 20.10.0.5430
```

### Workaround
A trivial solution is to make sure that both variables point to the same location. In this case everything works fine.

**Comment by tskir** [comment]
Hi @pditommaso! In light of the evidence I provided in my [previous comment](https://github.com/nextflow-io/nextflow/issues/1659#issuecomment-780291808), do you suppose we could reopen this ticket and remove the stale labels? The resolusion itself can of course wait as it's not a critical bug, but I suppose it could be still tracked as an open issue.

**Comment by RosaDeSa** [comment]
I'm facing the same error.
Everything worked in my case until singularity was updated to singularity ce version 4.0.3 on my HPC and I cleaned the singularity cache.

I defined in .bashrc file the export SINGULARITY_CACHEDIR="/home/r.desantis/singularity/cachedir"
and export NXF_SINGULARITY_CACHEDIR="/home/r.desantis/singularity/cachedir"

However, when I try to pull a docker container with singularity in my nextflow pipeline, I have this error

Command error:
ERROR: Unknown image format/type: /home/users/scratch/work/singularity/staphb-seqtk.img

In the singularity cachedir, I found the container with the .img extension, but it seems singularity doesn't recognize it.

Before having the unknown image error, I have this message during the pipeline running
`Pulling Singularity image docker://staphb/seqtk:1.3 [cache /home/r.desantis/singularity/cachedir/staphb-seqtk-1.3.img]. `
So it seems that nextflow pulls the singularity container in exactly the directory that I defined in .bashrc.

If I manually pull the image with the current version of singularity, and run it, everything works:

```
singularity --version
singularity-ce version 4.0.3
singularity pull docker://staphb/seqtk:1.3
INFO:    Using cached SIF image
singularity run seqtk_1.3.sif
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Singularity> 

```
It seems like if I pull manually the container, singularity creates the .sif image instead if I pull the image inside the pipeline, singularity pull the image with the extension .img

I've also update the nextflow version to 23.10.1. 
And I declared in my nextflow.config 

```
   singularity {
       enabled    = true
       autoMounts = true
}
```
